app/service/rpc_service.ts

```diff
@@ -186,18 +186,46 @@ class RpcService {
         init.body = lengthPrefixMessage(requestData);
 
         const reader = (await this.fetch(url, "stream", init))?.getReader();
+        let buffer = new Uint8Array();
+        let bufferedLength = 0;
+        let totalLength = 0;
         while (true) {
           let result = await reader?.read();
           if (result?.done) {
             return;
           }
 
-          // Ignore the first 5 byte length-prefix, see the comment for lengthPrefixMessage for more info.
-          // TODO(siggisim): Support messages that come across flushes, though I'm not sure if this will
-          // happen much (if at all) in practice.
-          let value = result?.value.slice(5) || new Uint8Array();
-          callback(null, value);
+          // Skip empty results so we don't have to deal with undefined values below.
+          if (!result?.value) {
+            continue;
+          }
+
+          // If we haven't seen a message yet, create a new buffer sized based on the length prefix.
+          if (totalLength == 0) {
+            totalLength = new DataView(result.value.buffer, 1, 4).getUint32(0, false /* big endian */);
+            let message = result.value.slice(5);
+            buffer = new Uint8Array(totalLength);
+            buffer.set(message);
+            bufferedLength = message.length;
+          } else if (bufferedLength < totalLength) {
+            // If we already have a buffer, just tack onto it.
+            buffer.set(result.value, bufferedLength);
```

Comment:
This `buffer.set` will throw a `RangeError` if the server can return a chunk that contains the end of one message and the start of another (i.e. crossing chunk boundaries) - will that ever happen or do we always flush before sending a new chunk?

(I think `RangeError` is fine if we're not expecting it to happen but maybe worth adding a comment)

Reply:
Yeah, after some more testing - I managed to hit that case. Got a fix in progress.

Reply:
sweet - just as a heads up you might also hit the case where the length prefix crosses chunk boundaries - e.g. only 3 of the expected 4 bytes - I remember having to deal with that in my hackathon project haha

Reply:
I rewrote things a bit to make them a bit more readable, and added support for leftovers (i.e. the first case where we have the start of the next message, or even multiple full messages and a chunk) and split length prefixes.

Took a few attempts to get something that felt good, but I'm not unhappy with where it ended up. LMK what you think.

Reply:
Looks good! The `readAtLeastNBytes` helps a lot